src/api/c/random_engine.cpp

```diff
@@ -0,0 +1,357 @@
+        RandomEngine out;
+
+        out.type = engine.type;
+        out.seed = engine.seed;
+        out.counter = engine.counter;
+        AF_CHECK(af_retain_array(&out.pos, engine.pos));
+        AF_CHECK(af_retain_array(&out.sh1, engine.sh1));
+        AF_CHECK(af_retain_array(&out.sh2, engine.sh2));
+        out.mask = engine.mask;
+        AF_CHECK(af_retain_array(&out.recursion_table, engine.recursion_table));
+        AF_CHECK(af_retain_array(&out.temper_table, engine.temper_table));
+        AF_CHECK(af_retain_array(&out.state, engine.state));
+
+        *outHandle = getRandomEngineHandle(out);
+
+    } CATCHALL;
+    return AF_SUCCESS;
+}
+
+af_err af_random_engine_set_type(af_random_engine *engine, const af_random_type rtype)
+{
+    try {
+        RandomEngine e = *(getRandomEngine(engine));
+        if (rtype != e.type) {
+            bool empty;
+            if (rtype == AF_RANDOM_MERSENNE_GP11213) {
+                af_is_empty(&empty, e.state);
+                if (empty) {
+                    AF_CHECK(af_create_array(&e.pos, pos, 1, &MaxBlocks, u32));
+                    AF_CHECK(af_create_array(&e.sh1, sh1, 1, &MaxBlocks, u32));
+                    AF_CHECK(af_create_array(&e.sh2, sh2, 1, &MaxBlocks, u32));
+                    e.mask = mask;
+                    AF_CHECK(af_create_array(&e.recursion_table, recursion_tbl, 1, &TableLength, u32));
+                    AF_CHECK(af_create_array(&e.temper_table, temper_tbl, 1, &TableLength, u32));
+                    e.state = getHandle(initMersenneState(e.seed, getArray<uint>(e.recursion_table)));
+                }
+            }
+            e.type = rtype;
+        }
+    } CATCHALL;
+    return AF_SUCCESS;
+}
+
+af_err af_default_random_engine_set_type(const af_random_type rtype)
+{
+    try {
+        af_random_engine e = DefaultRandomEngine();
+        af_random_engine_set_type(&e, rtype);
+    } CATCHALL;
+    return AF_SUCCESS;
+}
+
+af_err af_random_engine_set_seed(af_random_engine *engine, const uintl seed)
+{
+    try {
+        AF_CHECK(af_init());
+        RandomEngine *e = getRandomEngine(*engine);
+        e->seed = seed;
+        if (e->type == AF_RANDOM_MERSENNE_GP11213) {
+            initMersenneState(getWritableArray<uint>(e->state), seed, getArray<uint>(e->recursion_table));
+        } else {
+            e->counter = 0;
+        }
+    }
+    CATCHALL;
+    return AF_SUCCESS;
+}
+
+af_err af_random_engine_get_seed(uintl * const seed, af_random_engine engine)
+{
+    try {
+        AF_CHECK(af_init());
+        RandomEngine *e = getRandomEngine(engine);
+        *seed = e->seed;
+    }
+    CATCHALL;
+    return AF_SUCCESS;
+}
+
+af_err af_random_engine_uniform(af_array *out, af_random_engine engine, const unsigned ndims, const dim_t * const dims, const af_dtype type)
+{
+    try {
+        af_array result;
+        AF_CHECK(af_init());
+
+        af::dim4 d = verifyDims(ndims, dims);
+        RandomEngine *e = getRandomEngine(engine);
```

Comment:
Again, if there is way to verify the `engine` is valid or not, it should be done here.